### PR TITLE
New Listening Mode + some bug fixes and cleanup

### PR DESCRIPTION
**Pull request by digant73**
**IMPROVEMENTS:**
* Added TFT Listening Mode: The TFT simply listen for ACK messages (e.g. spontaneous) while all the outgoing messages are purged.
This mode is useful on main boards sharing the bus with TFT and USB (e.g. MKS GEN L) allowing Octoprint/Pronterface to properly control the printer (via USB) avoiding command collisions (otherwise present with TFT in normal mode).
The TFT will receive the ACK messages (spontaneous and/or related to gcodes from USB port) displaying on the screen all the status info.
The mode can be activated/deactivated on "Connection" menu. Existing labels (ON/OFF) and icon have been used to avoid to add new ones

**NOTE:** The mode state is stored on FLASH allowing to restore the mode at startup. This allows to power on/off the printer remotely and to control the printer via USB without the need of any touch (from the user) on the display to engage the mode
* Moved some functions from commnet.c/h to Printing.c/h API
* Minor code reduction
* Removed some global variables
* Minor cleanup

**BUG FIXES:**
* Fixed a bug preventing the TFT to reconnect to printer when switching from Marlin mode to Touch mode

Connection_1

Connection_2

Ready_1

### CODE REVIEW

File: Copy to SD Card root directory to update/config.ini [hunk 1]
```diff
@@ -824,16 +824,16 @@ end_gcode_enabled:0
 cancel_gcode_enabled:0
 
 #### Start Gcode
-# This gcode will runs before starting a print if `start_gcode_enabled` is enabled.
+# This gcode will runs before starting a print if "start_gcode_enabled" is enabled.
```
Comment (kisslorand) [inline Copy to SD Card root directory to update/config.ini:827]: ~runs~ - run

File: Copy to SD Card root directory to update/config.ini [hunk 2]
```diff
@@ -824,16 +824,16 @@ end_gcode_enabled:0
...
 #   Value range: [min: 3, max: 75 characters]
 start_gcode:G28 XY R20\n
 
 #### End Gcode
-# This gcode will runs after a print is completed if `end_gcode_enabled` is enabled.
+# This gcode will runs after a print is completed if "end_gcode_enabled" is enabled.
```
Comment (kisslorand) [inline Copy to SD Card root directory to update/config.ini:832]: ~runs~ - run

File: Copy to SD Card root directory to update/config.ini [hunk 3]
```diff
@@ -824,16 +824,16 @@ end_gcode_enabled:0
...
 #   Value range: [min: 3, max: 75 characters]
 end_gcode:M104 S0\nM140 S0\nM107\nM18\n
 
 #### Cancel Gcode
-# This gcode will runs when a print is canceled if `cancel_gcode_enabled` is enabled.
+# This gcode will runs when a print is canceled if "cancel_gcode_enabled" is enabled.
```
Comment (kisslorand) [inline Copy to SD Card root directory to update/config.ini:837]: ~runs~ - run

File: Copy to SD Card root directory to update/config_rrf.ini [hunk 4]
```diff
@@ -819,16 +819,16 @@ end_gcode_enabled:0
 cancel_gcode_enabled:0
 
 #### Start Gcode
-# This gcode will runs before starting a print if `start_gcode_enabled` is enabled.
+# This gcode will runs before starting a print if "start_gcode_enabled" is enabled.
```
Comment (kisslorand) [inline Copy to SD Card root directory to update/config_rrf.ini:822]: ~runs~ - run

File: Copy to SD Card root directory to update/config_rrf.ini [hunk 5]
```diff
@@ -819,16 +819,16 @@ end_gcode_enabled:0
...
 #   Value range: [min: 3, max: 75 characters]
 start_gcode:G28 XY R20\n
 
 #### End Gcode
-# This gcode will runs after a print is completed if `end_gcode_enabled` is enabled.
+# This gcode will runs after a print is completed if "end_gcode_enabled" is enabled.
```
Comment (kisslorand) [inline Copy to SD Card root directory to update/config_rrf.ini:827]: ~runs~ - run

File: Copy to SD Card root directory to update/config_rrf.ini [hunk 6]
```diff
@@ -819,16 +819,16 @@ end_gcode_enabled:0
...
 #   Value range: [min: 3, max: 75 characters]
 end_gcode:M104 S0\nM140 S0\nM107\nM18\n
 
 #### Cancel Gcode
-# This gcode will runs when a print is canceled if `cancel_gcode_enabled` is enabled.
+# This gcode will runs when a print is canceled if "cancel_gcode_enabled" is enabled.
```
Comment (kisslorand) [inline Copy to SD Card root directory to update/config_rrf.ini:832]: ~runs~ - run

File: TFT/src/User/API/Settings.c [hunk 7]
```diff
@@ -214,34 +221,29 @@ void setupMachine(void)
...
     infoMachineSettings.longFilename = infoSettings.long_filename;
 
-  mustStoreCmd("M503 S0\n");
+  if (infoMachineSettings.firmwareType == FW_SMOOTHIEWARE)  // Smoothieware does not report detailed M115 capabilities
+  {
+    infoMachineSettings.leveling        = BL_ABL;
+    infoMachineSettings.emergencyParser = ENABLED;
+  }
 
   if (infoMachineSettings.firmwareType == FW_REPRAPFW)
```
Comment (kisslorand) [inline TFT/src/User/API/Settings.c:238]: I would use "else if" for redundant "if" statements

File: TFT/src/User/Menu/PrintingMenu.c [hunk 8]
```diff
@@ -187,12 +187,10 @@ void menuBeforePrinting(void)
   // initialize things before the print starts
   progDisplayType = infoSettings.prog_disp_type;
   layerDisplayType = infoSettings.layer_disp_type * 2;
-  setLayerNumber(0);
```
Comment (kisslorand) [inline TFT/src/User/Menu/PrintingMenu.c:190]: This was a part of a bugfix. Why do you removed it?

File: TFT/src/User/Menu/PrintingMenu.c [hunk 9]
```diff
@@ -187,12 +187,10 @@ void menuBeforePrinting(void)
...
   coordinateSetAxisActual(Z_AXIS, 0);
   coordinateSetAxisTarget(Z_AXIS, 0);
   setM73_presence(false);
-  setTotalTime(0);
```
Comment (kisslorand) [inline TFT/src/User/Menu/PrintingMenu.c:194]: Why did you remove this?

**Comment by kisslorand** [comment]
I do not agree with the things moved out from comment.c/h. In my book it is not an optimization creating new procedures/functions and calling them instead of direct accessing a variable. 
Those variables were used only in conjunction with comment info so I still think they belong to where they were.

File: Copy to SD Card root directory to update/config_rrf.ini (hunk 6 above)
Comment (digant73) [reply Copy to SD Card root directory to update/config_rrf.ini:832]: yes

File: TFT/src/User/API/Settings.c (hunk 7 above)
Comment (digant73) [reply TFT/src/User/API/Settings.c:238]: yes

File: TFT/src/User/Menu/PrintingMenu.c (hunk 8 above)
Comment (digant73) [reply TFT/src/User/Menu/PrintingMenu.c:190]: because "layer" and "total time" were moved to "infoPrinting" data struct that is initialized to 0 in printStart() function

File: TFT/src/User/Menu/PrintingMenu.c (hunk 9 above)
Comment (digant73) [reply TFT/src/User/Menu/PrintingMenu.c:194]: see comment above

**Comment by digant73** [comment]
@kisslorand but they are printing information used by Printing menu that doesn't need to know/interface with other modules (comment). Layer, time could be set by other modules/gcode while Printing menu simply retrieve and display them

**Comment by kisslorand** [comment]
Now that you removed M73 Rxx check, what happens if it is present?

**Comment by digant73** [comment]
it was always set to "false" in the code unless there is a case where it is set to "true" that I didn't see

File: TFT/src/User/API/interfaceCmd.c [hunk 10]
```diff
@@ -609,12 +618,11 @@ void sendQueueCmd(void)
           if (cmd_seen('R'))
           {
             setPrintRemainingTime((cmd_value() * 60));
-            setM73_presence(0);  // disable parsing remaning time from gCode comments
```
Comment (kisslorand) [inline TFT/src/User/API/interfaceCmd.c:612]: Oopsie! Here should be `setM73_presence(true)`.
I do not know what happened here because I never use numeric values for boolean, always `true` or `false`.

Comment (kisslorand) [reply TFT/src/User/API/interfaceCmd.c:612]: > it was always set to "false" in the code unless there is a case where it is set to "true" that I didn't see

See my comment in "interface.c" line 612. You just discovered a bug that I cannot explain how it slipped there.

Comment (digant73) [reply TFT/src/User/API/interfaceCmd.c:612]: ok. reverted and fixed

Comment (kisslorand) [reply TFT/src/User/API/interfaceCmd.c:612]: Thank you. Cheers!

**Comment by digant73** [comment]
@guruathwal testing this PR I found a bug on Terminal. It seems that some texts will cause an infinite loop on the "while loop" (at line 787 in Terminal.c). The menuTerminalWindow menu will start an infinite page update (from 1 to 20) with no text displayed on screen.
I found that the following text is one causing the infinite loop:
//action:notification Bed Cooling...

The loop is due to the presence of both "//" and ":". With theese characters, the exit condition on the loop is never met (or probably something inside the loop broke the code (e.g. getCharacterInfo)).

When you have some time, could you please have a look to that issue? it seems quite important.

Another minor bug I saw scrolling up on the window is that the ack messages (that were dispalyed with white color) are then displayed as a request command (orange color).

**Comment by guruathwal** [comment]
@digant73  Thank you for notifying me. I will look into it now.

**Comment by gordo3di** [comment]
I've been having random communication timeouts when using Repetier server connected via USB. Do you think this issue is related?

**Comment by digant73** [comment]
@gordo3di no, I think it's not related to that PR. This PR simply purges (skips) all the gcodes that should be sent to the printer
